python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,446 @@
+            lambda v: mmh3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        TimestamptzType: (
+            lambda v: mmh3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        LongType: (
+            lambda v: mmh3.hash(struct.pack("q", v)),
+            lambda t: t in [LongType, TimeType, TimestampType, TimestamptzType],
+        ),
+        StringType: (
+            lambda v: mmh3.hash(v),
+            lambda t: t == StringType,
+        ),
+        BinaryType: (
+            lambda v: mmh3.hash(v),
+            lambda t: t == BinaryType,
+        ),
+        UUIDType: (
+            lambda v: mmh3.hash(
+                struct.pack(
+                    ">QQ",
+                    (v.int >> 64) & 0xFFFFFFFFFFFFFFFF,
+                    v.int & 0xFFFFFFFFFFFFFFFF,
+                )
+            ),
+            lambda t: t == UUIDType,
+        ),
+        # bucketing by Float/Double is not allowed by the spec, but they have hash implementation
+        FloatType: (
+            lambda v: mmh3.hash(struct.pack("d", v)),
+            lambda t: t == FloatType,
+        ),
+        DoubleType: (
+            lambda v: mmh3.hash(struct.pack("d", v)),
+            lambda t: t == DoubleType,
+        ),
+    }
+
+    def __init__(self, source_type: Type, num_buckets: int):
+        if (
+            source_type not in Bucket._FUNCTIONS_MAP
+            and not isinstance(source_type, FixedType)
+            and not isinstance(source_type, DecimalType)
+        ):
+            raise ValueError(f"Cannot bucket by type: {source_type}")
+
+        super().__init__(
+            f"bucket[{num_buckets}]",
+            f"transforms.bucket(source_type={repr(source_type)}, num_buckets={num_buckets})",
+        )
+        self._type = source_type
+        self._num_buckets = num_buckets
+
+        if isinstance(self._type, FixedType):
+            self._hash_func = lambda v: mmh3.hash(v)
+            self._can_transform = lambda t: isinstance(t, FixedType)
+        elif isinstance(self._type, DecimalType):
+            self._hash_func = lambda v: mmh3.hash(transform_util.decimal_to_bytes(v))
+            self._can_transform = lambda t: isinstance(t, DecimalType)
+        else:
+            self._hash_func = Bucket._FUNCTIONS_MAP[self._type][0]
+            self._can_transform = Bucket._FUNCTIONS_MAP[self._type][1]
+
+    @property
+    def num_buckets(self) -> int:
+        return self._num_buckets
+
+    def apply(self, value) -> Optional[int]:
+        if value is None:
+            return None
+
+        return (self._hash_func(value) & Bucket._MAX_32_BITS_INT) % self._num_buckets
+
+    def can_transform(self, target: Type) -> bool:
+        return self._can_transform(target)
+
+    def result_type(self, source: Type):
+        return IntegerType
+
+
+class Time(Transform):
```

Comment:
This class name isn't very clear. Since we can't hide these classes like we do in Java, we should probably make sure that the names are distinctive, like `TimeTransform`.

Reply:
SG